pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,189 @@
+		}
+	}
+
+	sortByFinishedTime(newSucceededPods, currentTime)
+	sortByFinishedTime(newFailedPods, currentTime)
+	newBackoffInfo := backoffInfo{
+		key,
+		oldBackoffInfo.failuresAfterLastSuccess,
+		oldBackoffInfo.lastFailureTime,
+	}
+
+	if len(newSucceededPods) == 0 && len(newFailedPods) == 0 {
+		return newBackoffInfo
+	}
+
+	if len(newFailedPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = 0
+		newBackoffInfo.lastFailureTime = nil
+		return newBackoffInfo
+	}
+
+	if len(newSucceededPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + int32(len(newFailedPods))
+		lastFailureTime := getFinishedTime(newFailedPods[len(newFailedPods)-1], currentTime)
+		newBackoffInfo.lastFailureTime = &lastFailureTime
+		return newBackoffInfo
+	}
+
+	newBackoffInfo.failuresAfterLastSuccess = 0
+	newBackoffInfo.lastFailureTime = nil
+
+	lastSuccessTime := getFinishedTime(newSucceededPods[len(newSucceededPods)-1], currentTime)
+	for i := len(newFailedPods) - 1; i >= 0; i-- {
+		failedTime := getFinishedTime(newFailedPods[i], currentTime)
+		if lastSuccessTime.After(failedTime.Time) || lastSuccessTime.Equal(&failedTime) {
+			break
+		}
+		if newBackoffInfo.lastFailureTime == nil {
+			newBackoffInfo.lastFailureTime = &failedTime
+		}
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + 1
+	}
+
+	return newBackoffInfo
+}
+
+func sortByFinishedTime(pods []*v1.Pod, currentTime metav1.Time) {
+	finishedTimeSortFunction := func(p1, p2 *v1.Pod) bool {
+		p1FinishTime := getFinishedTime(p1, currentTime)
+		p2FinishTime := getFinishedTime(p2, currentTime)
+
+		return p1FinishTime.Before(&p2FinishTime)
+	}
+
+	sort.Slice(pods, func(i, j int) bool {
+		return finishedTimeSortFunction(pods[i], pods[j])
+	})
+}
+
+func getFinishedTime(p *v1.Pod, currentTime metav1.Time) metav1.Time {
+	var finishTime *metav1.Time
+	for _, containerState := range p.Status.ContainerStatuses {
+		if containerState.State.Terminated == nil {
+			finishTime = nil
+			break
+		}
+
+		if finishTime == nil {
+			finishTime = &containerState.State.Terminated.FinishedAt
+		} else {
+			if finishTime.Before(&containerState.State.Terminated.FinishedAt) {
+				finishTime = &containerState.State.Terminated.FinishedAt
+			}
+		}
+	}
+
+	if finishTime == nil {
+		return currentTime
+	}
+
+	return *finishTime
+}
+
+func getRemainingBackoffTime(backoff backoffInfo, clock clock.WithTicker, defaultBackoff time.Duration, maxBackoff time.Duration) time.Duration {
+	if backoff.failuresAfterLastSuccess == 0 {
+		return 0 * time.Second
+	}
+
+	f := int32(math.Ceil(1 + math.Log2(float64(maxBackoff.Seconds())/float64(defaultBackoff.Seconds()))))
+	var backoffInSeconds int64
+	if backoff.failuresAfterLastSuccess >= f {
+		backoffInSeconds = int64(maxBackoff.Seconds())
+	} else {
+		backoffInSeconds = int64(float64(defaultBackoff.Seconds()) * math.Pow(2, float64(backoff.failuresAfterLastSuccess-1)))
+	}
```

Comment:
this is overly complicated. I think you can just calculate what is the time we should sync, then subtract the current time

Reply:
- The way that is being used now is: `backOff - (Now() - lastFailureTime)`
- The way you suggest is: `(lastFailureTime + backOff) - Now()

Is my understanding correct?

Reply:
I would say:
- avoid recomputing things which are constant, such as `.Seconds()` or `f`. Maybe we can declare constants `MaxBackOffSeconds` and `DefaultBackoffSeconds`
- rename `f` to something meaningful, `MaxBackoffFailuresCount` and precompute.
- one way to simplify the lower part would be:
```golang
	if remainingBackoff := time.Duration(backoffInSeconds)*time.Second - clock.Since(*backoff.lastFailureTime); remainingBackoff >= 0 {
		return remainingBackoff
	}
	return 0
```
here, you could actually avoid also the set of computing `backoffInSeconds` and directly compute `backoff`.

Reply:
Note that you are first doing a Log2 and later a Pow, because you are trying to compare number of failures. If you compare times instead, there will be less math involved.

Reply:
Simplified the logic using simple for loop. Please let me know if this looks fine.

Reply:
Resolving. Please re-open if required.